Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +181,145 @@ def upgrade_formula(f)
       nil
     end
   end
+
+  def upgradable_dependents(kegs, formulae)
+    formulae_to_upgrade = {}
+    formulae_pinned = {}
+
+    formulae.each do |formula|
+      dependents_to_upgrade = []
+
+      dependents = kegs.select do |k|
+        k.runtime_dependencies
+         .any? { |d| d["full_name"] == formula.full_name }
+      end
+
+      next if dependents.empty?
+
+      dependents.each do |k|
+        f = Formulary.from_keg(k)
+        next unless f.outdated?(fetch_head: ARGV.fetch_head?)
+        next if formulae_to_upgrade[f.full_name]
+        next if formulae_pinned[f.full_name]
+
+        if f.pinned?
+          formulae_pinned[f.full_name] = f
+        else
+          dependents_to_upgrade << f
+        end
+      end
+
+      u, p = upgradable_dependents(kegs, dependents_to_upgrade)
+
+      formulae_to_upgrade.merge u
+      formulae_pinned.merge p
+    end
+
+    [formulae_to_upgrade, formulae_pinned]
+  end
+
+  def broken_dependents(kegs, formulae)
+    formulae_to_reinstall = {}
```

Comment:
Might be worth making this a `Set` instead; `Formulary` does its own caching and it'll ease correct merging.